### ISSUE TITLE
Cyclocomp Linter for R6 classes

### ISSUE THREAD

**Issue by pakjiddat** [r-lib/lintr#804]
I get the following message when running lintr on a file containing a R6 class:

**"functions should have cyclomatic complexity of less than 25, this has 33."**

The line number points to the class definition. I think instead of pointing to the class, it needs to point to the specific function which has high cyclomatic complexity.

**Comment by AshesITR** [comment]
This is caused by running `cyclocomp` on the top-level expressions (that is on the call to `R6Class()` in this case).

**Comment by kpagacz** [comment]
Is the solution to skip running cyclocomp on calls to R6Class?

**Comment by AshesITR** [comment]
Not really, since that wouldn't lint overly complex R6 member functions.
I think the key is to obtain better location information for the lint, optimally by running on all function definitions separately.

**Comment by MichaelChirico** [comment]
I guess this is actually a cyclocomp bug.

it doesn't make sense IMO to run cyclocomp on the class definition, but on the constituent methods.

e.g., I guess I would be impossible to get an R6 class passing if it has a sufficient number of methods. is cyclocomp telling us it's bad to have a lot of methods in a class?

**Comment by IndrajeetPatil** [comment]
I think a concrete example should help this discussion.

I am linting here `processx::process` R class; deliberately with a lower limit so that we can see that message.

Can anyone see how `{cyclocomp}` comes with complexity of `7`? 
I just see one `if()` statement in the destructor, and don't understand what is contributing to complexity here.

``` r
library(lintr)

lint(
  text = 'process <- R6::R6Class(
  "process",
  cloneable = FALSE,
  public = list(
    initialize = function(command = NULL, args = character(),
                          stdin = NULL, stdout = NULL, stderr = NULL, pty = FALSE,
                          pty_options = list(), connections = list(), poll_connection = NULL,
                          env = NULL, cleanup = TRUE, cleanup_tree = FALSE, wd = NULL,
                          echo_cmd = FALSE, supervise = FALSE, windows_verbatim_args = FALSE,
                          windows_hide_window = FALSE, windows_detached_process = !cleanup,
                          encoding = "", post_process = NULL) {
      process_initialize(
        self, private, command, args, stdin,
        stdout, stderr, pty, pty_options, connections,
        poll_connection, env, cleanup, cleanup_tree, wd,
        echo_cmd, supervise, windows_verbatim_args,
        windows_hide_window, windows_detached_process,
        encoding, post_process
      )
    },
    finalize = function() {
      if (!is.null(private$tree_id) && private$cleanup_tree &&
        ps::ps_is_supported()) {
        self$kill_tree()
      }
    },
    kill = function(grace = 0.1, close_connections = TRUE) {
      process_kill(self, private, grace, close_connections)
    },
    kill_tree = function(grace = 0.1, close_connections = TRUE) {
      process_kill_tree(self, private, grace, close_connections)
    },
    signal = function(signal) {
      process_signal(self, private, signal)
    },
    interrupt = function() {
      process_interrupt(self, private)
    },
    get_pid = function() {
      process_get_pid(self, private)
    },
    is_alive = function() {
      process_is_alive(self, private)
    },
    wait = function(timeout = -1) {
      process_wait(self, private, timeout)
    },
    get_exit_status = function() {
      process_get_exit_status(self, private)
    },
    format = function() {
      process_format(self, private)
    },
    print = function() {
      process_print(self, private)
    },
    get_start_time = function() {
      process_get_start_time(self, private)
    },
    is_supervised = function() {
      process_is_supervised(self, private)
    },
    supervise = function(status) {
      process_supervise(self, private, status)
    },
    read_output = function(n = -1) {
      process_read_output(self, private, n)
    },
    read_error = function(n = -1) {
      process_read_error(self, private, n)
    },
    read_output_lines = function(n = -1) {
      process_read_output_lines(self, private, n)
    },
    read_error_lines = function(n = -1) {
      process_read_error_lines(self, private, n)
    },
    is_incomplete_output = function() {
      process_is_incompelete_output(self, private)
    },
    is_incomplete_error = function() {
      process_is_incompelete_error(self, private)
    },
    has_input_connection = function() {
      process_has_input_connection(self, private)
    },
    has_output_connection = function() {
      process_has_output_connection(self, private)
    },
    has_error_connection = function() {
      process_has_error_connection(self, private)
    },
    has_poll_connection = function() {
      process_has_poll_connection(self, private)
    },
    get_input_connection = function() {
      process_get_input_connection(self, private)
    },
    get_output_connection = function() {
      process_get_output_connection(self, private)
    },
    get_error_connection = function() {
      process_get_error_connection(self, private)
    },
    read_all_output = function() {
      process_read_all_output(self, private)
    },
    read_all_error = function() {
      process_read_all_error(self, private)
    },
    read_all_output_lines = function() {
      process_read_all_output_lines(self, private)
    },
    read_all_error_lines = function() {
      process_read_all_error_lines(self, private)
    },
    write_input = function(str, sep = "\n") {
      process_write_input(self, private, str, sep)
    },
    get_input_file = function() {
      process_get_input_file(self, private)
    },
    get_output_file = function() {
      process_get_output_file(self, private)
    },
    get_error_file = function() {
      process_get_error_file(self, private)
    },
    poll_io = function(timeout) {
      process_poll_io(self, private, timeout)
    },
    get_poll_connection = function() {
      process_get_poll_connection(self, private)
    },
    get_result = function() {
      process_get_result(self, private)
    },
    as_ps_handle = function() {
      process_as_ps_handle(self, private)
    },
    get_name = function() {
      ps_method(ps::ps_name, self)
    },
    get_exe = function() {
      ps_method(ps::ps_exe, self)
    },
    get_cmdline = function() {
      ps_method(ps::ps_cmdline, self)
    },
    get_status = function() {
      ps_method(ps::ps_status, self)
    },
    get_username = function() {
      ps_method(ps::ps_username, self)
    },
    get_wd = function() {
      ps_method(ps::ps_cwd, self)
    },
    get_cpu_times = function() {
      ps_method(ps::ps_cpu_times, self)
    },
    get_memory_info = function() {
      ps_method(ps::ps_memory_info, self)
    },
    suspend = function() {
      ps_method(ps::ps_suspend, self)
    },
    resume = function() {
      ps_method(ps::ps_resume, self)
    }
  ),
  private = list(
    command = NULL, 
    args = NULL,
    cleanup = NULL, # cleanup argument
    cleanup_tree = NULL, # cleanup_tree argument
    stdin = NULL, # stdin argument or stream
    stdout = NULL, # stdout argument or stream
    stderr = NULL, # stderr argument or stream
    pty = NULL, # whether we should create a PTY
    pty_options = NULL, # various PTY options
    pstdin = NULL, # the original stdin argument
    pstdout = NULL, # the original stdout argument
    pstderr = NULL, # the original stderr argument
    cleanfiles = NULL, # which temp stdout/stderr file(s) to clean up
    wd = NULL, # working directory (or NULL for current)
    starttime = NULL, # timestamp of start
    echo_cmd = NULL, # whether to echo the command
    windows_verbatim_args = NULL,
    windows_hide_window = NULL,
    status = NULL, # C file handle
    supervised = FALSE, # Whether process is tracked by supervisor
    stdin_pipe = NULL,
    stdout_pipe = NULL,
    stderr_pipe = NULL,
    poll_pipe = NULL,
    encoding = "",
    env = NULL,
    connections = list(),
    post_process = NULL,
    post_process_result = NULL,
    post_process_done = FALSE,
    tree_id = NULL,
    get_short_name = function() {
      process_get_short_name(self, private)
    },
    close_connections = function() {
      process_close_connections(self, private)
    }
  )
)',
linters = cyclocomp_linter(complexity_limit = 5L))
#> <text>:1:1: style: [cyclocomp_linter] Functions should have cyclomatic complexity of less than 5, this has 7.
#> process <- R6::R6Class(
#> ^
```

<sup>Created on 2022-09-22 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by kpagacz** [comment]
You might have better luck finding an answer to your question in this [repo](https://github.com/MangoTheCat/cyclocomp). `lintr` uses the `cyclocomp` package for this particular feature. But maybe some of the guys here are also knowledgeable enough about `cyclocomp` to help you out.

As a wild guess, the number of member functions impacts the complexity.

**Comment by IndrajeetPatil** [comment]
> it doesn't make sense IMO to run cyclocomp on the class definition, but on the constituent methods.

I agree with @MichaelChirico here.

> I guess this is actually a cyclocomp bug.

I agree that this issue is better discussed and addressed in `{cyclocomp}` itself, and then `{lintr}` will adjust itself downstream accordingly.
